### PR TITLE
eclass: Add support for dev-util/samurai

### PR DESCRIPTION
**Pull request by orbea**
This adds support for using `dev-util/samurai` instead of `dev-util/ninja` which works almost everywhere except for a few cases that need not yet implemented features (i.e. https://github.com/michaelforney/samurai/issues/36).

This can be controlled by the user setting `NINJA=ninja` (Default) or `NINJA=samu` in their environment and can be forced in an ebuild by setting the variable before inheriting any eclasses.

There are three cases which should work:

* both samurai and ninja are installed and the user might use either
* only samurai is installed
* only ninja is installed

The last two commits get rid of hardcoded instances of `ninja` in `sys-devel/llvm` and `sys-devel/clang`.

### CODE REVIEW

File: eclass/cmake.eclass
Comment (mgorny) [inline eclass/cmake.eclass:709]:
```suggestion
	local generator=${CMAKE_MAKEFILE_GENERATOR}
	case ${CMAKE_MAKEFILE_GENERATOR} in
		ninja) generator="eninja" ;;
	esac
```

File: eclass/meson.eclass
Comment (mgorny) [inline eclass/meson.eclass:51]:
```suggestion
	${NINJA_DEPEND}
```

You're dropping the outer quotes here.

File: eclass/ninja-utils.eclass
Comment (mgorny) [inline eclass/ninja-utils.eclass:51]: If you let the user override it, you're violating the metadata cache immutability rule; i.e. if the users set `NINJA=samu` (which I suppose we want to allow), they will have different deps than users who don't.

File: sys-devel/clang/clang-12.0.1.ebuild
Comment (mgorny) [inline sys-devel/clang/clang-12.0.1.ebuild:133]: Why do you need the fallback value?

**Review by mgorny — CHANGES_REQUESTED** [review]

File: eclass/ninja-utils.eclass
Comment (mgorny) [reply eclass/ninja-utils.eclass:51]: Scratch that. I see cmake is doing something like that anyway, so *shrug*.

**Review by thesamesam — APPROVED** [review]
Apart from mgorny's comments, I'm happy, and this is essentially what I had locally after iterating on your patches (to add the choice element).

Thank you!

File: sys-devel/clang/clang-12.0.1.ebuild
Comment (orbea) [reply sys-devel/clang/clang-12.0.1.ebuild]: I was trying to be overly cautious, but yea its pointless and I removed the fallback.

File: eclass/cmake.eclass
Comment (orbea) [reply eclass/cmake.eclass]: That is indeed cleaner, I changed it. Thanks!

**Comment by orbea** [comment]
I made the changes as suggested, thanks for the review!

File: eclass/ninja-utils.eclass
Comment (orbea) [reply eclass/ninja-utils.eclass:51]: I was unhappy with that too, but I'm not yet familiar enough with the eclasses to improve this. My original patch didn't allow for setting `NINJA` in the ebuild itself, but that was suggested by @thesamesam on irc which is how I ended up here.

**Comment by orbea** [comment]
@mgorny Is there anything else I can do to improve this?

**Review by mgorny — APPROVED** [review]
Nah, I think it's good enough to get mgorny's stamp of approval™.

**Comment by thesamesam** [comment]
Sent to ML.

**Comment by orbea** [comment]
@mgorny I updated the cmake.eclass commit with your suggestion from the ML, it is indeed seems like a far better idea.

> I'm sorry for missing this earlier but could we perhaps reuse
> cmake_build here?  Instead of this whole block above, something like:
>
> DESTDIR="${D}" cmake_build install "$@"

File: eclass/cmake.eclass
Comment (floppym) [inline eclass/cmake.eclass:368]: I would just drop this check. BDEPEND should suffice to ensure that ninja or samurai is installed.

File: eclass/meson.eclass
Comment (floppym) [inline eclass/meson.eclass:288]: I think we should just do this in global scope, within the case statement.

File: eclass/cmake.eclass
Comment (orbea) [reply eclass/cmake.eclass]: Sure, I removed it.

File: eclass/meson.eclass
Comment (orbea) [reply eclass/meson.eclass:288]: I tried this at first, but it spammed the warning many times while resolving dependencies. I'm not sure there are any examples of `ewarn` used in the global scope without an explicit error?

**Comment by floppym** [comment]
I pushed the ninja-utils and meson changes. I will leave the other changes to be merged by their respective maintainers.

**Review by floppym — APPROVED** [review]

**Comment by orbea** [comment]
Thanks, I really appreciate it! I rebased the PR against master as well.

File: eclass/cmake.eclass
Comment (thesamesam) [inline eclass/cmake.eclass:368]: Why are we dropping this again?

Comment (orbea) [reply eclass/cmake.eclass:368]: I received this reply on the ML from Mike Gilbert.
```
>  	# this could happen if ninja is manually enabled (eg. make.conf) but not installed  
-	if [[ ${CMAKE_MAKEFILE_GENERATOR} == ninja ]] && ! has_version -b dev-util/ninja; then
+	if [[ ${CMAKE_MAKEFILE_GENERATOR} == ninja ]] && [[ -n "${NINJA_DEPEND}" ]] && ! has_version -b "${NINJA_DEPEND}"; then

I would just drop this check. BDEPEND should suffice to ensure that ninja or samurai is installed.
```
I can add it back if desired, it doesn't really matter to me.

Comment (thesamesam) [reply eclass/cmake.eclass:368]: Oh, sorry, I missed that. No, go wild.

**Comment by orbea** [comment]
After the ninja and meson changes were merged into the master branch emerge now wants to pull in ninja as a dependency despite `NINJA="samu"` in my `make.conf`. However when adding `echo`  to the case statement in `ninja-utils.eclass` to confirm its selecting samurai it begins to work as expected. Removing the `echo` and now it fails again. I don't understand what is going on?

**Comment by thesamesam** [comment]
> After the ninja and meson changes were merged into the master branch emerge now wants to pull in ninja as a dependency despite `NINJA="samu"` in my `make.conf`. However when adding `echo` to the case statement in `ninja-utils.eclass` to confirm its selecting samurai it begins to work as expected. Removing the `echo` and now it fails again. I don't understand what is going on?

I suspect it's to do with https://github.com/gentoo/gentoo/pull/25272#discussion_r862430276

**Comment by orbea** [comment]
> I suspect it's to do with [#25272 (comment)](https://github.com/gentoo/gentoo/pull/25272#discussion_r862430276)

I had the same suspicion, but I really don't understand why the workaround works.

**Comment by floppym** [comment]
The metadata cache that is distributed via rsync does not have NINJA set in the environment, so it will always contain the dev-util/ninja dependency.

Portage uses the MD5 hash of the eclass to determine if the cache is valid. Making any change to the eclass code after syncing will invalidate the metadata cache.

In other words, if you invalidate the cache by making a change to ninja-utils.eclass, portage will re-compute the dependency, taking your NINJA setting in make.conf into account.

**Comment by orbea** [comment]
Thanks for explaining, but I'm not sure what should be done about this? The original patches I sent to the ML a while ago set it to use either ninja or samurai.
```
 NINJA_DEPEND="|| ( >=dev-util/ninja-1.8.2 dev-util/samurai )"
```
This didn't have this problem, but prevented the variable from being able to set in the ebuild itself as suggested on irc. To be clear being able to force the ninja implementation in the ebuild itself has limited value currently and will only fail in a few places for samurai for niche ninja features that have not been implemented yet.

Still it seems problematic that there is not a clear way to allow user controllable variables that can be overridden by the ebuild or am I just missing something?

I also imagine there is a similar less obvious issue with `CMAKE_MAKEFILE_GENERATOR` in `cmake.eclass`.

**Comment by mgorny** [comment]
You don't do anything about it. If the user wants to force samu via `make.conf`, then the user needs to install it manually. If the user wants to get rid of ninja, `package.provided`, I guess.

**Comment by orbea** [comment]
> f the user wants to get rid of ninja, package.provided, I guess.

Ah thanks, I didn't know about that.

File: eclass/cmake.eclass
Comment (a17r) [reply eclass/cmake.eclass:368]: That snippet was introduced to cmake-utils.eclass by commit e9cd2c48c36ac18a3b50ab09e1ced545260d4b30.

It helps covering for Portage failing to build things in the right order: https://bugs.gentoo.org/561570

Comment (floppym) [reply eclass/cmake.eclass:368]: Bug [561570](https://bugs.gentoo.org/561570) was caused by toralf setting CMAKE_MAKEFILE_GENERATOR in make.conf, not by some bug in Portage.

Setting CMAKE_MAKEFILE_GENERATOR in make.conf violates the metadata invariance rule, and users must ensure they install the necessary packages manually in that case.

I don't think this has_version check in the eclass is wanted or necessary.

**Pull request closed without merging**
